crates/compiler/parse/src/remove_spaces.rs

```diff
@@ -598,12 +598,103 @@ impl<'a> RemoveSpaces<'a> for StrLiteral<'a> {
     fn remove_spaces(&self, arena: &'a Bump) -> Self {
         match *self {
             StrLiteral::PlainLine(t) => StrLiteral::PlainLine(t),
-            StrLiteral::Line(t) => StrLiteral::Line(t.remove_spaces(arena)),
-            StrLiteral::Block(t) => StrLiteral::Block(t.remove_spaces(arena)),
+            StrLiteral::Line(t) => {
+                let mut needs_merge = false;
+                let mut last_was_mergable = false;
+                for segment in t.iter() {
+                    let mergable = matches!(
+                        segment,
+                        StrSegment::Plaintext(_)
+                            | StrSegment::Unicode(_)
+                            | StrSegment::EscapedChar(_)
+                    );
+                    if mergable && last_was_mergable {
+                        needs_merge = true;
+                        break;
+                    }
+                    last_was_mergable = mergable;
+                }
+
+                if !needs_merge {
+                    return StrLiteral::Line(t.remove_spaces(arena));
+                }
+
+                let mut new_segments = Vec::new_in(arena);
+                let mut last_text = String::new_in(arena);
+
+                normalize_str_segments(arena, t, &mut last_text, &mut new_segments);
+                if !last_text.is_empty() {
+                    new_segments.push(StrSegment::Plaintext(last_text.into_bump_str()));
+                }
+
+                StrLiteral::Line(new_segments.into_bump_slice())
+            }
+            StrLiteral::Block(t) => {
+                let mut new_segments = Vec::new_in(arena);
+                let mut last_text = String::new_in(arena);
+                for line in t {
+                    normalize_str_segments(arena, line, &mut last_text, &mut new_segments);
+                }
+                if !last_text.is_empty() {
+                    new_segments.push(StrSegment::Plaintext(last_text.into_bump_str()));
+                }
+
+                StrLiteral::Line(new_segments.into_bump_slice())
+            }
         }
     }
 }
 
+fn normalize_str_segments<'a>(
+    arena: &'a Bump,
+    segments: &[StrSegment<'a>],
+    last_text: &mut String<'a>,
+    new_segments: &mut Vec<'a, StrSegment<'a>>,
+) {
+    for segment in segments.iter() {
+        match segment {
+            StrSegment::Plaintext(t) => {
+                last_text.push_str(t);
+            }
+            StrSegment::Unicode(t) => {
+                let hex_code: &str = t.value;
+                let c = char::from_u32(u32::from_str_radix(hex_code, 16).unwrap()).unwrap();
+                last_text.push(c);
```

Comment:
Should this be pushing more than just the u32? Like `\u(u32)` or something?

Reply:
That would work as well, but this guarantees the e.g. doing `"\u(41)"` gets normalized to `"a"`. However: doing it this way is simpler, and since this is in a part of the code that's essentially verifying other code (that also happens to _not_ have implications on user-visible output or other parts of the compiler), that simplicity seems important.